### DISCUSSION TITLE
Proof by computation

### DISCUSSION THREAD

**Discussion by parno** [verus-lang/verus#120] [Language design]
These notes summarize some discussions we've had regarding what "proof by computation" might look like in Verus.  Please feel free to add more ideas or discussion below.

Sometimes, we want to prove a property that ultimately boils down to computing a function over constants.  It could be something simple like `assert pow(2, 10) == 1024`, or something more complex, like these examples:
[nth_root_lemma](https://github.com/secure-foundations/veri-titan/blob/8ae4130652c5bf2adbcf8489239090f32d0608ec/crypto_examples/ntt_cleanup/omega.dfy#L99)
[CRC bit unrolling](https://github.com/vmware-labs/verified-betrfs/blob/ee4b18d553933440bb5ecda037c6a1c411a49a5f/lib/Crypto/CRC32LutBitUnrolling.i.dfy)
[CRC lut entries](https://github.com/vmware-labs/verified-betrfs/blob/ee4b18d553933440bb5ecda037c6a1c411a49a5f/lib/Crypto/CRC32Lut.i.dfy#L105)
[CRC lut powers](https://github.com/vmware-labs/verified-betrfs/blob/ee4b18d553933440bb5ecda037c6a1c411a49a5f/lib/Crypto/CRC32LutPowers.i.dfy)
These examples seem "obvious" to the developer, but it can be difficult to persuade the SMT solver to do the seemingly simple desired computation.

We've also seen examples, e.g., in F*, where partial evaluation can be useful, which requires support for "computing" over symbolic values too, not just over concrete constants.  Symbolic execution creates challenges, however, as a termination proof for a function only ensures termination on concrete values, while symbolic values may loop indefinitely.

To provide both concrete and symbolic computations, we would likely want a symbolic evaluator for VIR.  To simplify the user interface, our initial sense was that such evaluations would run subject to the same rlimit as the overall query, in the expectation that most evaluations would finish quickly or not at all.

Note that we still expect any postconditions of the computed function to be "learned" by the prover.  In other words, given an optimized function like:
```
#[exec]                                                           
fn calculate_powmodQ(base: nat, exp: nat) -> nat                    
{  
  ensures(|r| r == powmodQ(base, exp)
   ...
}
```
we could successfully write:
```
assert_by_compute(calculate_powmodQ(2, 1024) == 4);
assert(powmodQ(2, 1024) == 4);
```

One item we haven't discussed explicitly is how the programmer should invoke a proof-by-computation.  One possibility is the `assert_by_compute` used in the example above.  Another option would be to have a dedicated function, e.g., `compute`, so the programmer could write:
```
assert(compute(calculate_powmodQ(2, 1024)) == 4);
```
This would provide flexibility closer to that of F*'s `normalize` facility, which has been handy in the past (e.g., in [our work on Vale/F*](https://www.andrew.cmu.edu/user/bparno/papers/vale-popl.pdf) -- see bottom of page 12).

As a side note, a symbolic executor for VIR might also be useful for a future Verus debugger, in that it might allow us to respond to user queries about the values of variables or functions that were not defined in the counterexample returned by Z3.

**Comment by utaal** [comment]
Implemented by @parno in #275.
